### PR TITLE
[kai/caregiver-route] Add /caregiver route

### PR DESCRIPTION
**Pull request by kai-wei-mo**
## Notion ticket link

[Implement POST /caregiver route](https://www.notion.so/uwblueprintexecs/Implement-POST-caregiver-route-7670810e31e64c22bf7a350f32cfe025)

## Implementation description
* GET
  * gets all caregivers
  * if the caregiver table does not exist, returns `[]`
* POST
  * adds a caregiver
  * if the request is invalid (e.g. caretaker id already exists, foreign key not found, etc.), then do nothing
 

## Steps to test GET
1. `docker-compose down --volumes` (after this, the `caregivers` relation does not exist)
2. `docker-compose up`
3. `curl -X GET http://localhost:5000/caregiver` (`[]` is the chosen response when `caregivers` does not exist)
4. `docker exec -it cas_py_backend /bin/bash -c "flask db upgrade"`
5. `docker exec -it cas_py_backend /bin/bash -c "python -m tools.db_seed"`
6. `curl -X GET http://localhost:5000/caregiver` (check that there are 2 caregivers)

## Steps to test POST
- see *"What should reviewers focus on?"* before proceeding
1. `docker-compose down --volumes`
2. `docker-compose up`
3. `docker exec -it cas_py_backend /bin/bash -c "flask db upgrade"`
4. `docker exec -it cas_py_backend /bin/bash -c "python -m tools.db_seed"`
5. `curl -X POST -H "Content-Type: application/json" -d '{"type": "CAREGIVER", "first_name": "John", "last_name": "Doe", "is_primary": true, "child_id": 1, "address_id": 1, "relationship_to_child": "FOSTER_CAREGIVER", "phone_number": "1234567890"}' http://localhost:5000/caregiver/`

## What should reviewers focus on?
* question: the case where the `caretaker` table does not exist and we try to POST is not considered. this is because a valid `caretaker` creation needs a `children` foreign key. and if children exists, caretaker table should exist too. is this ok?
* when testing POST, the tester needs to run the `curl` command 3 times because the first and second runs do not see the existing 2 `caretaker`s created by the seed script. how to fix this?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by jennchenn** [comment]
just converting this as a draft for now -- feel free to change it to ready for review whenever!

File: backend/python/app/rest/__init__.py [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
 
-    app.register_blueprint(user_routes.blueprint)
-    app.register_blueprint(intake_routes.blueprint)
     app.register_blueprint(auth_routes.blueprint)
+    app.register_blueprint(caregiver_routes.blueprint)
+    app.register_blueprint(child_routes.blueprint)
     app.register_blueprint(documentation_routes.blueprint)
+    app.register_blueprint(intake_routes.blueprint)
+    app.register_blueprint(user_routes.blueprint)
     app.register_blueprint(visit_routes.blueprint)
```
Comment (kai-wei-mo) [inline backend/python/app/rest/__init__.py:18]: alphabetizing (i don't think that changing the order of registration breaks anything)

File: backend/python/app/rest/caregiver_routes.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+from ..services.implementations.caregiver_service import CaregiverService
+
+caregiver_service = CaregiverService(current_app.logger)
+
+blueprint = Blueprint("caregiver", __name__, url_prefix="/caregiver")
+
+
+# get all caregivers
+@blueprint.route("/", methods=["GET"], strict_slashes=False)
+# @require_authorization_by_role({"Admin"})
```
Comment (kai-wei-mo) [inline backend/python/app/rest/caregiver_routes.py:17]: should this be uncommented?

Comment (kai-wei-mo) [inline backend/python/app/rest/caregiver_routes.py:17]: should this be uncommented?

File: backend/python/app/rest/caregiver_routes.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        caregivers = caregiver_service.get_all_caregivers()
+        return jsonify(list(map(lambda user: user.__dict__, caregivers))), 200
+    except Exception as error:
+        return jsonify(error), 400
+
+
+# create a caregiver
+@blueprint.route("/", methods=["POST"], strict_slashes=False)
+# @require_authorization_by_role({"Admin"})
+# @validate_request("CreateCaregiverDTO")
```
Comment (kai-wei-mo) [inline backend/python/app/rest/caregiver_routes.py:29]: should this be uncommented?

File: backend/python/app/services/implementations/user_service.py [hunk 4]
```diff
@@ -112,7 +112,7 @@ def get_auth_id_by_user_id(self, user_id):
 
     def get_users(self):
         user_dtos = []
-        user_list = [result for result in User.query.all()]
```
Comment (kai-wei-mo) [inline backend/python/app/services/implementations/user_service.py:115]: i think this is a no-op

File: backend/python/app/rest/__init__.py (hunk 1 above)
Comment (kai-wei-mo) [inline backend/python/app/rest/__init__.py:18]: alphabetizing (i don't think changing the registration order breaks anything)

File: backend/python/tools/db_seed.py [hunk 5]
```diff
@@ -90,20 +98,31 @@ def insert_test_data():
...
+    db.engine.execute("TRUNCATE TABLE children RESTART IDENTITY CASCADE")
+    db.engine.execute("TRUNCATE TABLE children RESTART IDENTITY CASCADE")
     db.engine.execute("TRUNCATE TABLE concerns RESTART IDENTITY CASCADE")
+    db.engine.execute("TRUNCATE TABLE daytime_contacts RESTART IDENTITY CASCADE")
     db.engine.execute("TRUNCATE TABLE goals RESTART IDENTITY CASCADE")
     db.engine.execute("TRUNCATE TABLE intakes RESTART IDENTITY CASCADE")
-    db.engine.execute(
-        "TRUNCATE TABLE daytime_contacts RESTART IDENTITY CASCADE")
-    db.engine.execute("TRUNCATE TABLE children RESTART IDENTITY CASCADE")
+    db.engine.execute("TRUNCATE TABLE users RESTART IDENTITY CASCADE")
```
Comment (kai-wei-mo) [inline backend/python/tools/db_seed.py:123]: alphabetizing

File: backend/python/app/services/implementations/caregiver_service.py [hunk 6]
```diff
@@ -28,3 +28,15 @@ def create_caregiver(self, caregiver):
...
             self.logger.error(str(error))
             raise error
+
+    def get_all_caregivers(self):
+        if not db.engine.dialect.has_table(db.engine, "caregivers"):
+            return []
+
+        try:
+            caregivers = Caregiver.query.all()
+            caregivers_dto = [CaregiverDTO(**x.to_dict()) for x in caregivers]
```
Comment (jennchenn) [inline backend/python/app/services/implementations/caregiver_service.py:38]: nit
```suggestion
            caregivers_dto = [CaregiverDTO(**caregiver.to_dict()) for caregiver in caregivers]
```

File: backend/python/app/rest/caregiver_routes.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+# get all caregivers
+@blueprint.route("/", methods=["GET"], strict_slashes=False)
+# @require_authorization_by_role({"Admin"})
+def get_all_caregivers():
+    try:
+        caregivers = caregiver_service.get_all_caregivers()
+        return jsonify(list(map(lambda user: user.__dict__, caregivers))), 200
+    except Exception as error:
+        return jsonify(error), 400
```
Comment (jennchenn) [inline backend/python/app/rest/caregiver_routes.py:23]: i think we need this to be converted to `str` for the error to be json serializable

File: backend/python/app/rest/caregiver_routes.py (hunk 3 above)
Comment (jennchenn) [reply backend/python/app/rest/caregiver_routes.py:29]: yes - we can validate the request but keep `require_auth...` commented out for now

Comment (jennchenn) [reply backend/python/app/rest/caregiver_routes.py:29]: `validate_request` happens here: https://github.com/uwblueprint/childrens-aid-society/blob/main/backend/python/app/middlewares/validate.py

File: backend/python/app/services/implementations/caregiver_service.py [hunk 8]
```diff
@@ -28,3 +28,15 @@ def create_caregiver(self, caregiver):
             db.session.rollback()
             self.logger.error(str(error))
             raise error
+
+    def get_all_caregivers(self):
```
Comment (jennchenn) [inline backend/python/app/services/implementations/caregiver_service.py:32]: do you mind adding a comment here: `FIXME: change this to match spec for actual get caregivers method`

File: backend/python/tools/db_seed.py [hunk 9]
```diff
@@ -90,20 +98,31 @@ def insert_test_data():
...
 
-    # fmt: on
+    # Caregivers
+    values = [
+        (1, "CAREGIVER", "Yor", "Forger", True, 1, 1, "FOSTER_CAREGIVER", "1234567890"),
+        (2, "CAREGIVER", "Loid", "Forger", True, 1, 1, "FOSTER_CAREGIVER", "1234567890")
+    ]
+
+    for value in values:
+        insert_values(db, "caregivers", ("id", "type", "first_name", "last_name", "is_primary", "child_id", "address_id", "relationship_to_child", "phone_number", "cpin_number", "date_of_birth", "special_needs", "name_of_child", "kinship_worker_name", "kinship_worker_ext", "foster_care_coord_name", "foster_care_coord_ext", "limitations_for_access"), value)
```
Comment (jennchenn) [inline backend/python/tools/db_seed.py:108]: for all the inserts in this file, remove the `id` column

**Review by jennchenn — CHANGES_REQUESTED** [review]
> question: the case where the caretaker table does not exist and we try to POST is not considered. this is because a valid caretaker creation needs a children foreign key. and if children exists, caretaker table should exist too. is this ok?
yup this is fine, you can assume the tables exist

> when testing POST, the tester needs to run the curl command 3 times because the first and second runs do not see the existing 2 caretakers created by the seed script. how to fix this?
mentioned during call, but removing the hardcoded `id`s in the seed script seems to make this error go away!

File: backend/python/app/services/implementations/caregiver_service.py [hunk 10]
```diff
@@ -28,3 +28,18 @@ def create_caregiver(self, caregiver):
...
+        # FIXME: change this to match spec for actual get caregivers method
+        if not db.engine.dialect.has_table(db.engine, "caregivers"):
+            return []
+
```
Comment (jennchenn) [inline backend/python/app/services/implementations/caregiver_service.py:36]: nit: i think we are running the rest of the app under the assumption that the database is up and running (or else we will have a lot of errors everywhere else) so not sure if you need this

**Review by jennchenn — APPROVED** [review]
tested changes and it looks good! just one nit

**Pull request merged**